File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int n = s.size();
        if(n % 2 != 0)
            return false;

        stack<char> st;
        for(int i = 0; i<n; i++){
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
                st.push(s[i]);
            else{
                if(st.empty())
                    return false;
                else if(s[i] == ')' && st.top() == '(')
                    st.pop();
                else if(s[i] == '}' && st.top() == '{')
                    st.pop();
                else if(s[i] == ']' && st.top() == '[')
                    st.pop();
                else
                    return false;
            }
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        Deque<Character> dq = new ArrayDeque<Character>();
        Set<Character> openBraces = new HashSet<>();
        openBraces.add('(');
        openBraces.add('[');
        openBraces.add('{');
        for(Character ch : s.toCharArray()){
            if(dq.isEmpty() || openBraces.contains(ch)){
                dq.addLast(ch);
            }else if(!dq.isEmpty() && ((dq.peekLast() == '(' && ch == ')') || 
                    (dq.peekLast() == '[' && ch == ']') || (dq.peekLast() == '{' && ch == '}'))){
                dq.pollLast();
                //System.out.println("Hello");
            }else{
                return false;
            }
        }
        return dq.isEmpty();
    }
}